### PR TITLE
Added support for ConcurrentMap.computeIfPresent #11958

### PR DESCRIPTION
**Pull request by webashutosh**
The new `computeIfPresent()` implementation keeps a clone of the old/existing value and uses that for replace/delete operations. This takes care of the user modifying existing value in-place.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/IMap.java [hunk 1]
```diff
@@ -70,28 +70,6 @@
...
- *     return value;
- * });
- * </pre>
- * It should be replaced with:
- * <pre>
- * map.computeIfPresent("key", (key, value) -&gt; {
- *     return new ObjectWithSomeAttribute("newAttributeValue");
- * });
- * </pre>
- * </li>
```
Comment (webashutosh) [inline hazelcast/src/main/java/com/hazelcast/map/IMap.java]: Removing this part of the documentation as Hz is already compiled with Java 8, and we have a plan to support Java 8 methods (#14913). 
Happy to revert this change in case I should keep this until we actually address #14913

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java [hunk 2]
```diff
@@ -1025,4 +1026,27 @@ private static void checkNotPagingPredicate(Predicate predicate, String method)
...
+            V newValue = remappingFunction.apply(key, oldValueClone);
+            if (newValue != null) {
+                if (replaceInternal(key, oldValueData, toData(newValue))) {
+                    return newValue;
+                }
+            } else if (removeInternal(key, oldValueData)) {
+                return null;
+            }
+        }
+    }
```
Comment (webashutosh) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: I could not think of a way to combine the get() + replace() operation into a single remote call. Mainly because I don't know how to send the BiFunction code/body across the wire (the function may've used classes/utilities that are only present on the client JVM)

Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: I think this is main issue that needs to be resolved. The problem is that this implementation can consume a lot of resources because of the retrying and the long latency. It could even end up in a livelock.

The best approach is to get everything done in a single remote call. There is no retrying needed here because the operation will have exclusive access to the partition.

Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: Sure @pveentjer , I'd look into that then. 
Just to confirm - this means that we want the BiFunction also to execute on the remote machine, correct (because we need to know the existing value of the entry before we can apply the function)?

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: Yes, this probably will be the better approach. You're going to have to create a new operation which will contain the mapping function - similar to how the `EntryProcessor` is transmitted when you invoke `IMap#executeOnKey`. On that note, would it not be simpler if you wrapped the mapping function in an `EntryProcessor` and used the same methods that that method already uses?

Also, I don't see any client-side support (`ClientMapProxy`) or client tests. Please add those.

File: hazelcast/src/main/java/com/hazelcast/map/IMap.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/map/IMap.java]: Yes, let's remove that once we actually have the implementations ready and covered by tests.

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java (hunk 2 above)
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: Sounds good, thanks @mmedenjak , will do.

Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: @pveentjer @mmedenjak @taburet 
I tested scenarios where a client node (JVM A) would pass an `EntryProcessor` as an anonymous class or a lambda to a server node (JVM B). I can confirm that such calls fail on the sever node, because the server fails to deserialize the anonymous-class/lambda instance (the server node doesn't have that "class" on its classpath).

This may not be so much of a problem for `EntryProcessor`s (because people must be creating a named subclass file in advance, and must be deploying the subclass on both (client and server) JVMs.

However, when people call methods that take Java 8 functional classes as args, its very likely that they'd use a lambda (I've almost never seen anyone creating named subclass files for a `BiFunction` or a `Function`). In such cases, the `computeIfPresent()` will fail (since the server would fail to deserialize the lambda passed as the `BiFunction`).

Given this, should we continue evaluating the `BiFunction` on the client JVM itself? 
Otherwise the new/updated implementation of `computeIfPresent ()` would cause existing usages of the method to start failing (if someone is passing the `BiFunction` arg as a lambda)

Thoughts?
PS: Should I be discussing such things on Gitter instead?

Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: How about if you enable user code deployment? Does that work with lambdas? For 4.1, there's some work in progress to enable re-deployment of such classes using UCD.

Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: Hi @cangencer, great suggestion! I tested that if we enable UCD on both sides, and add all the client-only classes to the UCD config, then that works! 
There are some nuances to it, but I think I'd raise a PR using UCD and see where that goes.

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: You're right and there is another issue that we might have that we have thought of - classes might be defined on all instances but since `BiFunction` doesn't implement `Serializable` or any other serialization technique, it might not work until you either cast it or you define a named class with some serialisation technique defined. Unfortunately, UCD doesn't help there.

Feels like we should try to get the EP approach running but if it's too hacky, we might revert back to the original approach from running from the client. Yes, it has its' drawbacks but it's better than an infinite loop.

Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: Sure @mmedenjak - let me send through a PR with this approach (using EP + UCD) - like you said, we can take a decision then ...

**Comment by Holmistr** [comment]
Hi @webashutosh , is there anything we can help you out with to keep you going? :)

**Comment by webashutosh** [comment]
> Hi @webashutosh , is there anything we can help you out with to keep you going? :)

Hi @Holmistr, thanks for checking with me! All good for now, I'll soon be sending a PR (I was out on a holiday for a few days)

**Comment by webashutosh** [comment]
Just an update - I am coding-up an implementation which would execute the BiFunction remotely when possible (e.g. when the BiFunction is not a lambda). And locally otherwise. Having to test quite a few cases thats why its taking a bit too much time.

**Comment by webashutosh** [comment]
Hi @pveentjer @mmedenjak - I added a commit to provide alternate implementation. The javadoc provided in `IMap.java` would clarify the design. Can you please have a look?

File: hazelcast/src/test/java/com/hazelcast/map/ComputeConditionallyTests.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiFunction;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.fail;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class ComputeConditionallyTests extends HazelcastTestSupport {
```
Comment (webashutosh) [inline hazelcast/src/test/java/com/hazelcast/map/ComputeConditionallyTests.java]: These tests cover both the execution from both the client and the server (we need to execute the same set of tests for both). So I didn't write a separate test class for the client.

File: hazelcast/src/test/java/testsubjects/ClientOnlyStaticSerializableBiFunction.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+/*
```
Comment (webashutosh) [inline hazelcast/src/test/java/testsubjects/ClientOnlyStaticSerializableBiFunction.java]: From here down, its all test classes I wrote in non-hazelcast base package. 
These are to make sure that the solution works for static/non-static and serializable/non-serializable bi-functions

**Comment by webashutosh** [comment]
Hi @mmedenjak @pveentjer @Holmistr - just a gentle reminder .... this PR is ready for review.

**Comment by Holmistr** [comment]
Hi @webashutosh , we haven't forgot about you, don't worry :) The guys are a bit occupied right now, but we will try to do the review in the next days. Thank you for your patience and thank you even more for working on it!

File: hazelcast/src/test/java/testsubjects/NonStaticFunctionFactory.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.util.function.BiFunction;
+
+public class NonStaticFunctionFactory {
+
+    private NonStaticFunctionFactory() {
+    }
+
+    public static BiFunction<String, String, String> getAnonymousNonSerializableBiFunction(final String returnValue) {
+        return new BiFunction<String, String, String>() {
```
Comment (mmedenjak) [inline hazelcast/src/test/java/testsubjects/NonStaticFunctionFactory.java]: Maybe it's an overkill having a separate class for this, you can move this to `ComputeConditionallyTests`. 

Also, can you move the package `testsubjects` under `com.hazelcast.test`. And what does `testsubjects` stand for?

File: hazelcast/src/test/java/com/hazelcast/map/ComputeConditionallyTests.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        StaticNonSerializableBiFunction biFunction = new StaticNonSerializableBiFunction("new_value");
+        testComputeIfPresentForFunction(biFunction, "new_value");
+    }
+
+    @Test(expected = ClassNotFoundException.class)
+    public void testComputeIfPresentWithRemappingFunctionPresentOnClientJVMOnly() throws Throwable {
+        ClientOnlyStaticSerializableBiFunction biFunction = new ClientOnlyStaticSerializableBiFunction("new_value");
+        try {
+            testComputeIfPresentForFunction(biFunction, "new_value");
+        } catch (HazelcastSerializationException e) {
```
Comment (mmedenjak) [inline hazelcast/src/test/java/com/hazelcast/map/ComputeConditionallyTests.java]: You can use `com.hazelcast.internal.util.RootCauseMatcher` instead. See [here](https://github.com/hazelcast/hazelcast/blob/97ff57ba92dc1fbfa77720885e7db40757999080/hazelcast/src/test/java/com/hazelcast/client/impl/spi/impl/ClientInvocationFutureTest.java#L112-L119) for an example.

File: hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+package com.hazelcast.internal.util;
+
+import java.io.Serializable;
+
+public final class ClassMetadataUtil {
+
+    private ClassMetadataUtil() {
+    }
+
+    public static boolean isClassStaticAndSerializable(Class clazz) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java]: You can move these kind of methods in `SerializationUtil`. Specifically, there is already a similar method there - `SerializationUtil#checkSerializable`.

A bigger issue is that this does not cover some serialisation techniques that HZ supports, such as `IdentifiedDataSerializable`, `DataSerializable` and others. For instance (although it's not that common), the user can define a function like this:

```java
public class IDSBiFunction implements BiFunction<String, String, String>, IdentifiedDataSerializable {
    private String returnValue;
    public IDSBiFunction() {
    }

    public IDSBiFunction(String returnValue) {
        this.returnValue = returnValue;
    }

    @Override
    public String apply(String key, String oldValue) {
        return returnValue;
    }

    @Override
    public int getFactoryId() {
        return 1;
    }

    @Override
    public int getClassId() {
        return 0;
    }

    @Override
    public void writeData(ObjectDataOutput out) throws IOException {
        out.writeUTF(returnValue);
    }

    @Override
    public void readData(ObjectDataInput in) throws IOException {
        returnValue = in.readUTF();
    }
}
```

If this class is registered with a factory, such as: `config.getSerializationConfig().addDataSerializableFactory(1, typeId -> new IDSBiFunction());`, it will be possible to serialize and deserialize the class. 

Unfortunately, checking for all possible serialization strategies is probably not going to be easy to do in a utility method. I was initially thinking that we should simply try to serialise the function here but that might incur an unnecessary overhead for most users trying to submit non-serializable functions. So maybe for a best-effort approach we can additionally check if the class inherits `DataSerializable` or `Portable`.  If the class doesn't implement these but is serializable, we can always add more checks in the future.

@cangencer @taburet WDYT?

File: hazelcast/src/main/java/com/hazelcast/map/impl/BiFunctionExecutingEntryProcessor.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+package com.hazelcast.map.impl;
+
+import com.hazelcast.map.EntryProcessor;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+public class BiFunctionExecutingEntryProcessor<K, V> implements EntryProcessor<K, V, V> {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/BiFunctionExecutingEntryProcessor.java]: It would be great if you could make this class implement `IdentifiedDataSerializable` since currently this class uses java serialisation which is not efficient and does not allow class evolution. It does get complicated here, though. Once you add the interface, you'll have to implement some additional methods, like:

```
@Override
    public int getFactoryId() {
        return MapDataSerializerHook.F_ID;
    }

    @Override
    public int getClassId() {
        return MapDataSerializerHook.BI_FUNCTION_EXECUTING_EP;
    }

    @Override
    public void writeData(ObjectDataOutput out) throws IOException {
        out.writeObject(biFunction);
    }

    @Override
    public void readData(ObjectDataInput in) throws IOException {
        biFunction = in.readObject();
    }
```

This will also require adding a new row in `MapDataSerializerHook` but also a new client message and you'll need to modify the client protocol. I suggest you look at the `EntryRemovingProcessor` class and how it's instantiated and used from other parts of the code.

File: hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java (hunk 7 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java]: Talking with Marko, seems like the case where the provided instance would implement IDS is something very awkward and usually users would be ok with simple lambdas and `Serializable`. IDS and other serialization techniques would simply execute locally. That means you don't need to change anything. Does that sound ok @cangencer @taburet?

File: hazelcast/src/test/java/testsubjects/NonStaticFunctionFactory.java (hunk 5 above)
Comment (webashutosh) [reply hazelcast/src/test/java/testsubjects/NonStaticFunctionFactory.java]: **Overall background -** 
In `ComputeConditionallyTests`, ideally I wanted to test scenarios where the client and servers would be on different JVMs, and the client has classes that are NOT present on the server JVM.
To simulate this, I used a custom class-loader for the `server`. That way, the classes visible to the `client` aren't visible to the `server` classloader.

Further to this, I noticed that during class-loading, classes inside `com.hazelcast` have an additional fallback ([link](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/nio/ClassLoaderUtil.java#L244)). I wanted to avoid this fallback for the tests (to make the tests stricter)

**About naming the new package `testsubjects` -** 
I thought we can use this package to host classes that act like "subjects" of tests (classes that aren't tests themselves. Other tests use them like subjects). Happy to rename it ... suggestions?

**About the separate class - **
Yes, it's kind of an overkill. I wanted the anonymous class to have a package other than `com.hazelcast`. Any other solution comes to mind?

**Comment by mmedenjak** [comment]
Just to be clear, I like how this PR is looking and I think it's close to being done. Congrats @webashutosh ! We just need to polish some details now. Unfortunately turning the entry processor is going to be the most tedious part.

File: hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java (hunk 7 above)
Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java]: It should take BiFunctionEx instead of BiFunction. Then it would be automatically serializable.

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java]: @cangencer we're talking about overriding `java.util.concurrent.ConcurrentMap#computeIfPresent(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction)`. 

How would it accept `BiFunctionEx`  instead of `BiFunction`? Am I missing something?

Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java]: See the example I posted, it's possible to do it by adding another method and then upcasting. I would also suggest a "checkSerializable" method to fail fast, if something is not serializable: https://github.com/hazelcast/hazelcast-jet/blob/master/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java#L168

Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java]: distributedStream extends Stream -> DistributedStream is analogous to IMap and Stream is analogous to ConcurrentMap

**Comment by webashutosh** [comment]
> Just to be clear, I like how this PR is looking and I think it's close to being done. Congrats @webashutosh ! We just need to polish some details now. Unfortunately turning the entry processor to implement IDS is going to be the most tedious part.

Thanks @mmedenjak ! I'm more than happy to address all the comments, will do that soon. Thanks really for the detailed feedback and the encouragement :)

File: hazelcast/src/main/java/com/hazelcast/map/impl/BiFunctionExecutingEntryProcessor.java (hunk 8 above)
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/BiFunctionExecutingEntryProcessor.java]: Hi @mmedenjak , I am done with all changes. Just that I am not sure if/why would I need to add a new client message and modify the client protocol. I am using existing operations in the implementation for `ClientMapProxy.java`, which already have an existing implementation ... am I missing something?

File: hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java (hunk 7 above)
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java]: Hi @cangencer @mmedenjak ... I pushed a commit to address all feedback received so far. 
I didn't yet add a method to accept a BiFunctionEx. For `computeIfPresent()` implementation, in addition to checking if the BiFunction is Serializable, we also need to check if has a "static" class. So adding the new overloaded method would only have avoided the Serialization check. I thought the performance cost saving would be minimal (I am sure the JVM would also need to do a check or two to figure out which version of the overloaded `computeIfPresent()` to call).
Also, the `SerializationUtil.checkSerializable()` is more robust (in addition to checking Serialization interface, it actually tries to serialize the object, to detect non-serializable fields).
Does this sound okay? Happy to be corrected ...

File: hazelcast/src/main/java/com/hazelcast/map/impl/BiFunctionExecutingEntryProcessor.java (hunk 8 above)
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/BiFunctionExecutingEntryProcessor.java]: Also wanted to check, do I need to provide this implementation for clients in other languages too?

File: hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java (hunk 7 above)
Comment (cangencer) [reply hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java]: If you have `BiFunctionEx` it should automatically made serializable without having to cast it explicitly. That's the main benefit.

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java [hunk 9]
```diff
@@ -1025,4 +1028,41 @@ private static void checkNotPagingPredicate(Predicate predicate, String method)
...
         }
     }
+
+    @Override
+    public V computeIfPresent(K key,
+                              BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
+        checkNotNull(key, NULL_KEY_IS_NOT_ALLOWED);
+        checkNotNull(key, NULL_BIFUNCTION_IS_NOT_ALLOWED);
+
+        if (SerializationUtil.isClassStaticAndSerializable(remappingFunction)) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: IMHO Looks good now. You're going to have to add one more check. We have a feature calling "rolling upgrade" where a cluster can be consisted out of members with 1 minor version difference. This means, if this feature is released in 4.1, the cluster may contain members with version 4.0 which will not be able to deserialise the new entry processor. So, the check should be added:

```
Version clusterVersion = getNodeEngine().getClusterService().getClusterVersion();

        if (SerializationUtil.isClassStaticAndSerializable(remappingFunction)
                && clusterVersion.isGreaterOrEqual(Versions.V4_1)) {
```

File: hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java [hunk 10]
```diff
@@ -1984,4 +1988,39 @@ protected void onDestroy() {
             super.onDestroy();
         }
     }
+
+    @Override
+    public V computeIfPresent(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
+        checkNotNull(key, NULL_KEY_IS_NOT_ALLOWED);
+        checkNotNull(key, NULL_BIFUNCTION_IS_NOT_ALLOWED);
+
+        if (SerializationUtil.isClassStaticAndSerializable(remappingFunction)) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java]: This is where it gets tricky again. The client is not aware of the version of the cluster it is connected to. So it might happen that the function is serializable and static and is sent to the cluster but the cluster members fail to deserialise it. Unfortunately, this would mean there is no other option for the user but to not have a serializable function when calling from the client or to upgrade the member versions.

I don't have a good idea how to tackle this. As a poor-man's solution (but, still better than an infinite loop which we have now), we can avoid this optimisation on the client and simply use `computeIfPresentLocally` in all cases. WDYT? cc @sancar

File: hazelcast/src/test/java/testsubjects/NonStaticFunctionFactory.java (hunk 5 above)
Comment (mmedenjak) [reply hazelcast/src/test/java/testsubjects/NonStaticFunctionFactory.java]: Forgot about the `com.hazelcast` package rule. Thanks, let's keep it as it is.

File: hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java (hunk 7 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/util/ClassMetadataUtil.java]: I'm fine with the current state, resolving this comment. If @cangencer has other comments, let's open a new thread.

File: hazelcast/src/main/java/com/hazelcast/map/impl/BiFunctionExecutingEntryProcessor.java (hunk 8 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/map/impl/BiFunctionExecutingEntryProcessor.java]: Looks good to me. You don't need to provide an implementation for other languages.

**Comment by mmedenjak** [comment]
Great work @webashutosh ! From my point of view, you can add the cluster version checks and address the client-side comment and this PR is good to go.

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by Holmistr** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java (hunk 10 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java]: Hi, @mmedenjak  is right we need a fallback mechanism when server is not available. Having a fallback mechanism have greater scope. Lets make the `computeIfPresentLocally` only implementation for now for the client. The optimisation for the client can be addressed later.

File: hazelcast/src/test/java/com/hazelcast/map/ComputeConditionallyTests.java (hunk 3 above)
Comment (sancar) [reply hazelcast/src/test/java/com/hazelcast/map/ComputeConditionallyTests.java]: All the tests that are using clients should reside in client package.

Comment (sancar) [reply hazelcast/src/test/java/com/hazelcast/map/ComputeConditionallyTests.java]: I could not find the test runs `computeIfPresent` from the server. Is it missing ?

Comment (sancar) [reply hazelcast/src/test/java/com/hazelcast/map/ComputeConditionallyTests.java]: Sorry, I just saw them.

File: hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java (hunk 10 above)
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java]: Thanks @mmedenjak @sancar! I will push a commit to address all feedback.

**Comment by webashutosh** [comment]
Hi @mmedenjak @sancar , I've pushed changes to address all the feedback. Can you please have a look?

**Comment by Holmistr** [comment]
Just one thing, so we wouldn't forget about it - if @mmedenjak approves the PR, @webashutosh please, squash the commits before we merge it to keep the Git history clean :)

**Comment by mmedenjak** [comment]
Squashed using github "Squash and commit" :)

Thank you for the contribution @webashutosh !

**Comment by webashutosh** [comment]
Thank you so much @mmedenjak @sancar @Holmistr @pveentjer @cangencer ! Looking forward to making many more contributions :)

**Comment by Holmistr** [comment]
@webashutosh We're glad to have you! Actually, do you happen to have a Twitter account? I would like to tweet about your contribution as well :)

**Comment by webashutosh** [comment]
Hi @Holmistr , thanks so much! I am not active on social media at all actually ... I'd definitely request for a mention in case that changes :)

**Comment by dbrimley** [comment]
I love to see this.  Thanks @webashutosh
